### ISSUE TITLE
Multi-edit is lost on undo

### ISSUE THREAD

**Issue by smjgordon** [notepad-plus-plus/notepad-plus-plus#14480]
I'm getting to know the new multi-edit feature of Notepad++.  I've heard of it in a few other editors, but this is my first time experiencing it myself.  I would imagine that this is one of many of its teething problems.

### Description of the Issue
If I undo while using multi-edit, the multiple cursors I had active disappear, leaving only one cursor.

### Steps to Reproduce the Issue
1. Ctrl+Click in a few places, thereby activating multi-edit.
2. Type some characters.  Or delete some characters.  Or paste.  Whichever you fancy.
3. Undo (via menu or Ctrl+Z).
4. Redo (via menu or Ctrl+Y).

### Expected Behaviour
Keep the multiple cursors active throughout the steps.

### Actual Behaviour
At step 3, multi-edit is deactivated, leaving only one of the cursors active.  This is the last one by position in the document when the undone action is typing, and the first one by position when the undone action is deleting or pasting.
At step 4, the cursor that remains active becomes the other one (last versus first) relative to the step 3 behaviour.

### Debug Information
Notepad++ v8.6   (64-bit)
Build time : Nov 23 2023 - 16:58:44
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : "C:\Program Files\Notepad++\change.log" 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Home (64-bit)
OS Version : 22H2
OS Build : 19045.3693
Current ANSI codepage : 1252
Plugins : 
    ComparePlugin (2.0.2)
    mimeTools (2.9)
    NppConverter (4.5)
    NppExport (0.4)

**Comment by alankilborn** [comment]
My understanding of undo/redo is that only text operations are undone/redone.

**Comment by smjgordon** [comment]
What's that to do with anything?  This is about something that is happening in addition to operations being undone/redone and shouldn't be happening.

**Comment by alankilborn** [comment]
The way I read it, it sounds like you expect multi-caret changes to be tracked along with text changes through undo/redo steps.

**Comment by smjgordon** [comment]
This isn't a matter of tracking multi-caret changes specifically. When I undo, I expect the cursor position to revert to what it was immediately before the undone action. When there are multiple cursors, it goes without saying that this means the positions of all cursors. Practically speaking, if during an editing task involving multiple cursors I make a mistake and undo, I should be able to continue the overall task I was in the middle of without having to painstakingly locate and re-set all my cursors.On 13/12/2023, 11:27 alankilborn ***@***.***> wrote:

   
  The way I read it, it sounds like you expect multi-caret changes to be tracked along with text changes through undo/redo steps.

**Comment by alankilborn** [comment]
Good luck in getting your problem fixed.

**Comment by xomx** [comment]
@alankilborn 
> My understanding of undo/redo is that only text operations are undone/redone.

While I basically agree with you, I have also to say that e.g. MSVS behaves like the OP expects.
And it would be nice if the N++ could behave like that too (not saying that I will be involved in this).

**Comment by donho** [comment]
I have to admit that the concerning behaviour of Visual Studio is better than the one in Notepad++.
Unfortunately, this Scintilla behaviour is improbably overridden by Notepad++.

**Comment by smjgordon** [comment]
I think we're all overlooking an important question: What is the design of multi-edit?

From the way it behaves under most editing operations, it seems clear to me that the design is that it persists until cancelled by the user.  This can be done explicitly by pressing Esc, or by operations that naturally cancel it such as Select All, Ctrl+Home or Ctrl+End.  Undoing is neither of these.  I claim that this means it's a bug that undoing cancels the multi-edit.

> Unfortunately, this Scintilla behaviour is improbably overridden by Notepad++.

In other words, the bug is in Scintilla rather than in N++ itself?  How sure are you of this?  In that case, I suppose one of us needs to raise it in the Scintilla bug tracking system....

**Comment by xomx** [comment]
> the bug is in Scintilla rather than in N++ itself? How sure are you of this?

I tried Scintilla SciTE and the behavior is the same as in the N++.
